### ISSUE TITLE
How to repair a corrupted mp4 video?

### ISSUE THREAD

**Issue by sunshinetpu** [jcodec/jcodec#98]
Can I repair a corrupted mp4 video using Jcodec?

**Comment by svitvitskiy** [comment]
What codec is inside? h264 with aac?

Stan

**Comment by sunshinetpu** [comment]
@svitvitskiy Yes, it is h264 video with aac audio. Can you give me some advices?

**Comment by svitvitskiy** [comment]
If you are just looking to repair a couple of videos that you have I'd say
download some of the software (I know at least 2 apps that do that) or you
can do http://mp4repair.org/ . You would have to pay though.

There's no repair functionality built into JCodec but if you are looking to
build a repair app and you need JCodec as boilerplate I'll be happy to help.

Also let me know if this is something important for your business in which
case I might consider building something like that.

Stan

**Comment by sunshinetpu** [comment]
@svitvitskiy Yes, I want to create the app which allows to repair the corrupted mp4 videos. I guess we need to decode the corrupted video and change something in the metadata. But I don't know how to do it with JCodec.

**Comment by gordonpro** [comment]
I find a android app had implemented your requirement [mp4fix](https://play.google.com/store/apps/details?id=com.smamolot.mp4fix) . It without any NDK lib and I decompiled the app it obscured the logic of codes, but I can find the code of JCodec.

Fixing mp4 video with jcodec is achievable thing.

**Comment by gordonpro** [comment]
@svitvitskiy I need fix atom missed mp4 file. can you give me some tips.

**Comment by svitvitskiy** [comment]
Which atom is missing?

Could you post your file somewhere?

Stan

**Comment by gordonpro** [comment]
The scene is I use the phone as a camera to record video and then stop recording without normal close, it may be caused by low power auto poweroff.

I cannot play the mp4 file, ffmepg reports `moov atom not found.`
 


I can use the [mp4fix](https://play.google.com/store/apps/details?id=com.smamolot.mp4fix) to fix this problem.  it refers another file that recorded by same camera device and use same configurations.

[error.mp4](http://asdflk.oss-cn-hangzhou.aliyuncs.com/error.mp4)

**Comment by svitvitskiy** [comment]
Ok, here's the reason why mp4fix needs another file from the same device :

Basically what happens when you abnormally terminate the recording -- you
get your frame data written to a file with no header ( 'moov' atom ).

It is possible to 'hunt' for video and audio frames inside the data blob
that got written AND recover their positions. But other then that nothing
is known.

One crucial bit of information is codec initialization data -- SPS/PPS
structs for H264 and MPEG audio descriptor for AAC. These are normally
stored in moov/trak/mdia/minf/stbl/stsd/avc1/avcC and
moov/trak/mdia/minf/stbl/stsd/mp4a/esds.

There's no way to just 'make up' these structures as they contain all the
'choices' encoder made during the encoding process. The H.264 encoder has
about 10 important independent variables and many choices for the value of
those. Specific choices that the encoder make depend on the implementation
and settings.

The reason it is usually possible to 'borrow' these things from a file
being made on the same device is because hardware encoders are simplistic
and will usually make the same choices over and over again.

Now back to your project.

You could re-implement whatever mp4fix does. Some of JCodec's functions
would be a help in this case. It's not going to be trivial but it's
definitely doable.

If you want to avoid having to supply a similar file you would have to
maintain a library of all possible codec init segments and pick the one
based on some external indication, say input from a user.

Stan

**Comment by sunshinetpu** [comment]
The mp4fix app uses JAAD library for working with Mp4 file. There is another library called Mp4Parser, which provides a lot of class for working with mp4 file. But the main issue is how to find the position of video data and audio data. We have mdat atom of corrupted video and have a good video with the same parameters from the same device. How can we define the duration of tracks?
